store/store.go

```diff
@@ -33,60 +31,92 @@ The store can be assumed to contain all contiguous blocks between base and heigh
 */
 type BlockStore struct {
 	db dbm.DB
-
-	// mtx guards access to the struct fields listed below it. We rely on the database to enforce
-	// fine-grained concurrency control for its data, and thus this mutex does not apply to
-	// database contents. The only reason for keeping these fields in the struct is that the data
-	// can't efficiently be queried from the database since the key encoding we use is not
-	// lexicographically ordered (see https://github.com/tendermint/tendermint/issues/4567).
-	mtx    tmsync.RWMutex
-	base   int64
-	height int64
 }
 
 // NewBlockStore returns a new BlockStore with the given DB,
 // initialized to the last height that was committed to the DB.
 func NewBlockStore(db dbm.DB) *BlockStore {
-	bs := LoadBlockStoreState(db)
-	return &BlockStore{
-		base:   bs.Base,
-		height: bs.Height,
-		db:     db,
-	}
+	return &BlockStore{db}
 }
 
 // Base returns the first known contiguous block height, or 0 for empty block stores.
 func (bs *BlockStore) Base() int64 {
-	bs.mtx.RLock()
-	defer bs.mtx.RUnlock()
-	return bs.base
+	iter, err := bs.db.Iterator(
+		blockMetaKey(1),
+		blockMetaKey(1<<63-1),
+	)
+	if err != nil {
+		return 0
+	}
+	defer iter.Close()
+
+	if iter.Valid() {
```

Comment:
This needs to check for iterator errors too (after this block, since an errored iterator is invalid).

Reply:
Just to clear up my understanding, how should certain errors thrown by an iterator be handled:
```
it, err := bs.db.Iterator()  -> we should panic
err := it.Error() -> we should return an error if we can else ignore
err := it.Close() -> we should return an error if we can else ignore
```

Reply:
In general, all errors should be returned. In certain cases you may get away with ignoring close errors (e.g. defers whee you're sure the consequences are negligible). Panics should generally only be used when returning an error is very inconvenient and the only failure modes are programmer error.

Reply:
Okay, thanks. I will try address this all in a follow up PR